working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2304,19 +2372,47 @@ To match a pattern `p` against a value `v`:
         some `T` determined either by the pattern's explicit type argument or
         inferred from the matched value type.*
 
-    2.  If the length of the list determined by calling `length` is not equal to
-        the number of subpatterns, then the match fails. *This match failure
-        becomes a runtime exception if the list pattern is in an irrefutable
-        context.*
+    2.  Let `l` be the length of the list determined by calling `length` on `v`.
 
-    3.  Otherwise, for each element subpattern, in source order:
+    3.  Let `h` be the number of non-rest element subpatterns preceding the rest
+        element if there is one, or the number of subpatterns if there is no
+        rest element.
 
-        1.  Extract the element value `e` by calling `[]` on `v` with an
-            appropriate integer index.
+    4.  Let `t` be the number of non-rest element subpatterns following the rest
+        element if there is one, or zero otherwise.
 
-        2.  Match `e` against the element subpattern.
+    3.  If `p` has no rest element and `l` is not equal to `h` then the match
+        fails. If `p` has a rest element and `l` is less than `h + t` then the
+        match fails. *These match failures become runtime exceptions if the list
+        pattern is in an irrefutable context.*
 
-    4.  The match succeeds if all subpatterns match.
+    4.  Match the head elements. For `i` from `0` to `h - 1`, inclusive:
+
+        1.  Extract the element value `e` by calling `[]` on `v` with index `i`.
+
+        2.  Match the `i`th element subpattern against `e`.
+
+    5.  If there is a rest element and it has a subpattern:
+
+        1.  Let `r` be the result of calling `sublist()` on `v` with arguments
+            `h`, and `l - t`.
```

Comment:
We could do that, but it seems like a lot of new machinery to spec and implement for marginal benefit. I think just calling `sublist()` has a lot of value:

1. @leafpetersen pointed out that it preserves the runtime type of the original list, which could be useful.
2. It lets a custom `List` implementation control the returned type. For example, a `LoggingList` that logs every element access could return a sublist that also logs element access.
3. It's easy to explain to users because the semantics are defined in terms of an API they already know.